pkg/controller/cluster/cluster_controller.go

```diff
@@ -151,20 +204,99 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
-			return reconcile.Result{}, err
+			return reconcile.Result{}, errors.WithStack(err)
 		}
 
 		return reconcile.Result{}, nil
 	}
 
-	klog.Infof("reconciling cluster object %v triggers idempotent reconcile.", name)
+	klog.Infof("Reconciling cluster object %v triggers idempotent reconcile.", name)
 	if err := r.actuator.Reconcile(cluster); err != nil {
 		if requeueErr, ok := errors.Cause(err).(controllerError.HasRequeueAfterError); ok {
 			klog.Infof("Actuator returned requeue-after error: %v", requeueErr)
 			return reconcile.Result{Requeue: true, RequeueAfter: requeueErr.GetRequeueAfter()}, nil
 		}
 		klog.Errorf("Error reconciling cluster object %v; %v", name, err)
-		return reconcile.Result{}, err
+		return reconcile.Result{}, errors.WithStack(err)
 	}
 	return reconcile.Result{}, nil
 }
+
+// listChildren returns a list of Deployments, Sets, and Machines than have an ownerref to the given cluster
+func (r *ReconcileCluster) listChildren(ctx context.Context, cluster *clusterv1.Cluster) ([]runtime.Object, error) {
+	var children []runtime.Object
+
+	ns := cluster.GetNamespace()
+	opts := metav1.ListOptions{
+		LabelSelector: labels.FormatLabels(
+			map[string]string{clusterv1.MachineClusterLabelName: cluster.GetName()},
+		),
+	}
+
+	dfunc := func(ctx context.Context, m metav1.ListOptions) (runtime.Object, error) {
+		return r.clusterClient.MachineDeployments(ns).List(m)
+	}
+	sfunc := func(ctx context.Context, m metav1.ListOptions) (runtime.Object, error) {
+		return r.clusterClient.MachineSets(ns).List(m)
+	}
+	mfunc := func(ctx context.Context, m metav1.ListOptions) (runtime.Object, error) {
+		return r.clusterClient.Machines(ns).List(m)
+	}
+
+	deployments, err := pager.New(dfunc).List(ctx, opts)
+	if err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "failed to list MachineDeployments in %s for cluster %s", ns, cluster.Name)
+	}
+	dlist, ok := deployments.(*clusterv1.MachineDeploymentList)
+	if !ok {
+		return []runtime.Object{}, fmt.Errorf("Expected MachineDeploymentList, got %T", deployments)
+	}
+
+	sets, err := pager.New(sfunc).List(ctx, opts)
+	if err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "Failed to list MachineSets in %s", ns)
+	}
+	slist, ok := sets.(*clusterv1.MachineSetList)
+	if !ok {
+		return []runtime.Object{}, fmt.Errorf("Expected MachineSetList, got %T", sets)
+	}
+
+	machines, err := pager.New(mfunc).List(ctx, opts)
+	if err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "Failed to list MachineSets in %s", ns)
+	}
+	mlist, ok := machines.(*clusterv1.MachineList)
+	if !ok {
+		return []runtime.Object{}, fmt.Errorf("Expected MachineList, got %T", machines)
+	}
+
+	for _, d := range dlist.Items {
+		if pointsTo(&d.ObjectMeta, &cluster.ObjectMeta) {
+			children = append(children, d.DeepCopyObject())
+		}
+	}
+
+	for _, s := range slist.Items {
+		if pointsTo(&s.ObjectMeta, &cluster.ObjectMeta) {
+			children = append(children, s.DeepCopyObject())
```

Comment:
I wonder if we should only append MachineSets if they do **not** belong to a MachineDeployment?

Reply:
MachineSets created by a deployment are owned by that deployment, not the cluster: 

```yaml
    ownerReferences:
    - apiVersion: cluster.k8s.io/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: MachineDeployment
      name: whinnyapolis-machinedeployment
      uid: 44b5c291-ad90-11e9-b9d6-0242ac110003
    resourceVersion: "33872"
```

Reply:
So, there is potentially a bug in release-0.1 where the MachineSet controller is trying to add an ownerRef to the cluster, but it only does it if there are 0 ownerRefs. We identified similar logic in another controller either while triaging a bug, or working on alpha2 (I don't remember which), and I believe we decided that the length check for 0 ownerRefs was incorrect.

I think given this logic is in place in both the MachineDeployment and MachineSet controllers, your logic is ok. If, however, we ever change the logic and instead decide we want all MachineDeployments and MachineSets to have an ownerRef pointing back at the cluster, we should probably adjust your logic here so we don't add MachineSets that belong to MachineDeployments, or Machines that belong to MachineSets. Does that make sense?

Reply:
I've been thinking about this more and I think I need to revise my comment 😄.

tl;dr: we should ensure that all MachineDeployments, MachineSets, and Machines have a Cluster owner reference; otherwise, deleting a Cluster could miss deleting some objects, and we remain in this stuck state.

Longer version:

I think we may want to update the MachineDeployment controller to remove the length=0 check when adding the owner reference to the cluster. Otherwise, if for some reason a MachineDeployment already has an owner reference to something else (i.e. a user manually set it to something), the MachineDeployment will not have the cluster listed as an owner. When someone deletes the cluster, it won't trigger this logic to delete the MachineDeployment, which could give us the same result we have now - the cluster won't actually get deleted and things will be stuck running in circles.

The same comment applies to the MachineSet and Machine controllers.

WDYT @detiber @vincepri @chuckha